### PR TITLE
Redux dev: Split actionLog and consoleDispatch enhancers

### PR DESCRIPTION
**Pull request by sirreal**
## The issue

The console dispatcher redux store enhancers serves several purposes.

* It attaches the top-level redux store methods to the window object so they are accessible for development, including `dispatch` as its name implies
* It adds some logging for dispatched actions and exposes some operations on the logged actions by attaching them to the window object.

These are two separate concerns, and including them in a single store enhancer is not feasible:
* Console dispatching must happen near the top of the enhancer stack so that they are visible to other store enhancers, especially our app middleware. This is OK in the current implementation.
* Action logging must happen near the bottom of the enhancer stack so that it has visibility on the actions dispatch by other enhancers, especially our app middleware. This is broken in the current implementation, middleware dispatched actions are invisible to the console action logger.

## The proposed fix

This PR solves the issue with missing actions from `window.actionLog` by splitting the existing enhancer into two and placing the "logging" enhancer near the bottom of the stack:

* Console Dispatcher - This enhancer keeps its name and position at the top of the enhancer stack. It's implementation is vastly simplified as its only responsibility at this point is to expose the store methods on the `window` object if it exists.
* Action Logger - All code related to the action logger has moved from the Console Dispatcher into this enhancer. It has been added near the bottom of the enhancer stack, just before the enhancer for the browser extension redux devtools.

## Context

This issue was discovered while testing https://github.com/Automattic/wp-calypso/pull/20638#issuecomment-350991262

## Testing
1. `dispatch( { type: actionTypes.USER_PROFILE_LINKS_REQUEST } )`
1. Ensure network request fires as expected
1. `actionLog.filter( actionTypes.USER_PROFILE_LINKS_REQUEST )`
1. Ensure you see the request action as expected
1. `actionLog.filter( actionTypes.USER_PROFILE_LINKS_RECEIVE )`
1. Ensure you see the receive action as expected
1. Ensure other [redux store methods](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#store-methods) continue to behave as expected (no changes):
   * [`getState()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#getstate) / `state`: provides access to the current redux state
   * [`dispatch()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#dispatchaction): see above
   * [`subscribe()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#subscribelistener)
   * [`replaceReducer()`](https://github.com/reactjs/redux/blob/master/docs/api/Store.md#replacereducernextreducer)

You can try the steps mentioned in https://github.com/Automattic/wp-calypso/pull/20638#issuecomment-350991262 to ensure the fix works.

### CODE REVIEW

**Comment by sirreal** [comment]
> With this change, the data layer no longer issues a `fetch` when I use the console dispatcher - you can try it with:  `dispatch( { type: 'USER_PROFILE_LINKS_REQUEST' } )`

Thanks @tyxla! That makes total sense. We'd expect the `dispatch` part of this middleware to occur early, before the rest of the middlewares, and the `log` part of this to occur towards the end, after the rest of the middlewares. I'll have another look.

**Comment by dmsnell** [comment]
Thanks for working on this. We definitely need to establish logging later in the chain - maybe split these two bits apart from one another.

Also the title is a bit misleading and I'm sure the scope grew as you dug in to the issue here - might need updating :smile:

Do you think it might make sense to _move_ the code in one PR and then _change_ it in another? Then it would appear in the diff as basic renames/moves and then in the other PR as a small set of changes instead of as adding new files and deleting old ones. Right now it's confusing to separate the moves from the changes.

**Comment by sirreal** [comment]
> Do you think it might make sense to move the code in one PR and then change it in another?

After having another look at this, I'm not sure there would be two significant independent PRs. This PR really just moves the `actionLog` related code into another independent enhancer. Moving that code requires it to be added to the store enhancers, which is done in the correct position.

I've tried to avoid making significant changes. The only exceptions are:
* using rest/spread args for the enhancers which is what [Redux `applyMiddleware` does](https://github.com/reactjs/redux/blob/87071fd4ab71acc4fdd8b3db37d2d7ff08b724a3/src/applyMiddleware.js#L20-L21).
* [Invert the conditional](https://github.com/Automattic/wp-calypso/pull/20725/files#diff-b5287c8e411f9b84e6c5f9baab18fbcfR23) for attaching redux store methods to window. There are a small set of operations in the conditional block and the unconditional `return store` at the end seems clearer.

> the title is a bit misleading and I'm sure the scope grew as you dug in to the issue here

Good point, I've given the title and description a makeover, I hope the issue and intentions here are clear now.

File: client/state/console-dispatch/index.js [hunk 1]
```diff
@@ -1,100 +1,37 @@
...
-};
-
-export const consoleDispatcher = next => ( reducer, initialState ) => {
-	const store = next( reducer, initialState );
-
-	if ( 'undefined' === typeof window ) {
-		return store;
+		Object.defineProperty( window, 'state', {
+			enumerable: true,
+			get: () => store.getState(),
```
Comment (sirreal) [inline client/state/console-dispatch/index.js]: We should be able to drop the extra function here:

```js
{
  get: store.getState,
}
```

**Comment by dmsnell** [comment]
Thanks @sirreal - at first I didn't realize this was doing exactly what I was suggesting. The diff was confusing to me.

**Comment by sirreal** [comment]
Rebased to fix conflict.

**Comment by sirreal** [comment]
Rebased.

File: client/state/action-log/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * chain, passing straight through.
+ */
+
+const state = {
+	actionHistory: [],
+	shouldRecordActions: true,
+	historySize: 100,
+};
+
+export const queryToPredicate = query => {
```
Comment (tyxla) [inline client/state/action-log/index.js]: If we're moving this one, let's move along its tests from `state/console-dispatch/test/*` too

File: client/state/action-log/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+**A Redux [store enhancer](https://github.com/reactjs/redux/blob/master/docs/Glossary.md#store-enhancer)**
+
+The Action Logger is a Redux store enhancer which records the most recently dispatched redux action
+so they can be inspected in the development environment for debugging and interactive design.
+
+## Functionality
+
+### Action logging
+
+Like the [Redux DevTools](https://github.com/zalmoxisus/redux-devtools-extension), the console dispatcher can also track a history of the actions flowing through the store.
```
Comment (tyxla) [inline client/state/action-log/README.md]: I thought we're separating the concepts of `console dispatcher` and `action logger` here - if that's the case, should we replace `console dispatcher` with `action logger`?

File: client/state/action-log/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+This can be toggled and the size of the history log can be adjusted.
+
+The following examples represent commands and code actually run inside the browser console.
+These are not snippets of bundles JavaScript inside of Calypso.
```
Comment (tyxla) [inline client/state/action-log/README.md]: I know we're essentially copying this, but `snippets of bundles JavaScript` - something appears to be wrong here.

File: client/state/action-log/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+```js
+// start recording actions
+// enabled by default in development
```
Comment (tyxla) [inline client/state/action-log/README.md]: `in development` looks too broad, maybe alter it to `in the development environment`?

File: client/state/console-dispatch/README.md [hunk 6]
```diff
@@ -21,49 +21,10 @@ An alias to `getState()` has been provided for terseness in the console: `state`
...
-// the action types have also been exported
-// to the console for easy access to types
-// note that action types will benefit from
-// auto-complete in the console
-actionLog.filter( actionTypes.COMMENTS_LIKE )
-
-// pass a regex to filter matching action types
-actionLog.filter( /^USER/ )
-actionLog.filter( /receive|request$/i )
+// dispatch an action simply
```
Comment (tyxla) [inline client/state/console-dispatch/README.md]: > dispatch an action simply

Something doesn't seem right here with this sentence.

File: client/state/console-dispatch/README.md [hunk 7]
```diff
@@ -21,49 +21,10 @@ An alias to `getState()` has been provided for terseness in the console: `state`
...
+dispatch( { type: actionTypes.USER_SETTINGS_REQUEST } )
```
Comment (tyxla) [inline client/state/console-dispatch/README.md]: Do we want to also document where `actionTypes` is coming from?

File: client/state/action-log/index.js (hunk 2 above)
Comment (sirreal) [reply client/state/action-log/index.js]: Whoops bac2d5decf93063052a9c41f84bf6e877c91418d

**Pull request merged**